sympy/simplify/_cse_diff.py

```diff
@@ -0,0 +1,255 @@
+
+    Returns
+    =======
+
+    processed_replacements : list of (Symbol, expression) pairs
+        Processed replacement list, in the same format of the 'replacements' input list.
+
+    processed_reduced : list of SymPy expressions
+        Processed reduced list, in the same format of the 'reduced_expressions' input list.
+    """
+
+    def traverse(node, repl_dict):
+        if isinstance(node, Derivative):
+            return replace_all(node, repl_dict)
+        if not node.args:
+            return node
+        new_args = [traverse(arg, repl_dict) for arg in node.args]
+        return node.func(*new_args)
+
+    def replace_all(node, repl_dict):
+        result = node
+        while True:
+            free_symbols = result.free_symbols
+            symbols_dict = {k: repl_dict[k] for k in free_symbols if k in repl_dict}
+            if not symbols_dict:
+                break
+            result = result.xreplace(symbols_dict)
+        return result
+
+    repl_dict = dict(replacements)
+    processed_replacements = [(rep_sym, traverse(sub_exp, repl_dict)) for rep_sym, sub_exp in replacements]
+    processed_reduced = [red_exp.__class__([traverse(exp, repl_dict) for exp in red_exp]) for red_exp in reduced_expressions]
+
+    return processed_replacements, processed_reduced
+
+
+def _forward_jacobian_core(replacements, reduced_expr, wrt):
+    r"""
+    Core function to compute the Jacobian of an input Matrix of expressions through forward accumulation.
+    Takes directly the output of a CSE operation (replacements and reduced_expr), and an iterable
+    of variables (wrt) with respect to which to differentiate the reduced expression and returns the
+    Jacobian matrix in DAG form.
+
+    The function also returns a list of precomputed free symbols for each subexpression,
+    which are useful in the substitution process.
+
+    NOTE: The output Jacobian might not be always exactly in a DAG form, but close to it.
+
+    Parameters
+    ==========
+
+    replacements : list of (Symbol, expression) pairs
+        Replacement Symbols and relative Common Subexpressions that have been replaced during a CSE operation.
+
+    reduced_expr : list of SymPy expressions
+        The reduced expressions with all the replacements from the replacements list above.
+
+    wrt : Matrix, tuple, or list
+        Iterable of expressions with respect to which to compute the Jacobian Matrix
+
+    Returns
+    =======
+
+    replacements : list of (Symbol, expression) pairs
+        Replacement Symbols and relative Common Subexpressions that have been replaced during a CSE operation.
+        Compared to the input replacement list, the output one doesn't contain replacement symbols inside
+        Derivative's arguments.
+
+    jacobian : list of SymPy expressions
+        The list only contains one element, which is the Jacobian Matrix with elements
+        in reduced form (replacement symbols are present)
```

Comment:
```suggestion
        in reduced form (replacement symbols are present).
```

Reply:
Why is this a list and not simply the matrix?

Reply:
Also, does this return the same type of matrix as passed in?

Reply:
> Why is this a list and not simply the matrix?

cse always return the expression in a list